File: frontend/src/services/guild.ts
import { api } from './api';

export enum GuildRole {
    Member = 0,
    Officer = 1,
    Leader = 2
}

// Restoration of interfaces
export interface GuildMember {
    id: number;
    guildId: number;
    userId: number;
    user: {
        id: number;
        name: string;
        avatarUrl?: string;
        level: number;
    };
    role: GuildRole;
    joinedAt: string;
    contributionXP: number;
}

export interface Guild {
    id: number;
    name: string;
    description?: string;
    leaderId: number;
    level: number;
    xp: number;
    capacity: number;
    createdAt: string;
    isPrivate: boolean;
    inviteCode?: string;
    members: GuildMember[];
}

export interface CreateGuildDto {
    name: string;
    description?: string;
    isPrivate?: boolean;
    inviteCode?: string;
}

export const guildAPI = {
    search: async (query: string = ''): Promise<Guild[]> => {
        const lcQuery = query ? `?query=${encodeURIComponent(query)}` : '';
        const response = await api.get<Guild[]>(`/guilds${lcQuery}`);
        return response.data;
    },

    get: async (id: number): Promise<Guild> => {
        const response = await api.get<Guild>(`/guilds/${id}`);
        return response.data;
    },

    create: async (data: CreateGuildDto): Promise<Guild> => {
        const response = await api.post<Guild>('/guilds', data);
        return response.data;
    },

    join: async (id: number, inviteCode?: string): Promise<void> => {
        await api.post(`/guilds/${id}/join`, { inviteCode });
    },

    leave: async (): Promise<void> => {
        await api.post(`/guilds/leave`);
    }
};

File: backend/FocusArena.Domain/Entities/Guild.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace FocusArena.Domain.Entities;

public class Guild
{
    [Key]
    public int Id { get; set; }

    [Required]
    [MaxLength(50)]
    public string Name { get; set; } = string.Empty;

    public string? Description { get; set; }

    public int LeaderId { get; set; } // UserId of the leader
    [ForeignKey("LeaderId")]
    public virtual User? Leader { get; set; }

    public int Level { get; set; } = 1;

    public int XP { get; set; } = 0;

    public int Capacity { get; set; } = 10; // Default capacity

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    // Guild Access
    public bool IsPrivate { get; set; } = false;

    [MaxLength(20)]
    public string? InviteCode { get; set; }

    // Navigation
    public virtual ICollection<GuildMember> Members { get; set; } = new List<GuildMember>();
}
